artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/mirror/AMQPMirrorControllerSource.java

```diff
@@ -202,8 +241,35 @@ private static Properties getProperties(Message message) {
 
    @Override
    public void postAcknowledge(MessageReference ref, AckReason reason) throws Exception {
+
+      MirrorController targetController = getControllerTarget();
+
+      if (targetController != null || ref.getQueue() != null && (ref.getQueue().isInternalQueue() || ref.getQueue().isMirrorController())) {
+         if (logger.isTraceEnabled()) {
+            logger.trace(server + " rejecting postAcknowledge queue=" + ref.getQueue().getName() + ", ref=" + ref + " to avoid infinite loop with the mirror (reflection)");
+         }
+         return;
+      }
+
+      if (logger.isTraceEnabled()) {
+         logger.trace(server + " postAcknowledge " + ref);
+      }
+
       if (acks && !ref.getQueue().isMirrorController()) { // we don't call postACK on snfqueues, otherwise we would get infinite loop because of this feedback
-         Message message = createMessage(ref.getQueue().getAddress(), ref.getQueue().getName(), POST_ACK, ref.getMessage().getMessageID());
+         Long internalIDObject = (Long)ref.getMessage().getBrokerProperty(INTERNAL_ID_EXTRA_PROPERTY);
+         long internalID;
+         if (internalIDObject == null) {
+            internalID = ByteUtil.mixByteAndLong(localMirrorId, ref.getMessageID());
+         } else {
+            internalID = internalIDObject.longValue();
+            if (logger.isTraceEnabled()) {
+               logger.trace("server " + server + " acking message " + ref);
+            }
+         }
+         if (logger.isTraceEnabled()) {
+            logger.trace(server + " sending ack message from server " + ByteUtil.getFirstByte(internalID) + " with messageID=" + ByteUtil.removeFirstByte(internalID));
+         }
```

Comment:
I continue to think this is a bad idea.

In this bit of code we munge together two values to form one (perhaps dropping information from the original ID silently while doing so) unless that was already done before...then either way we take that value and un-munge it in order to log the two components. Then we send it over the wire, where the other side will unmunge it and hopefully come up with 2 values it understands, and end up doing a whole bunch of similar hoop jumping with them/it.

Meanwhile, on sending a regular message for delivery we use 2 entirely separate delivery annotations to carry the pristine values without any need to jump through hoops or possibly unwittingly throw away information from the ID. Always transmitting the values separately and simplifying the whole situation just seems like a far better way to go.

Reply:
you might be acknowleding a message that was sent from another node.

Node1 sent the message, you failed over to 2, you could be acking a message that was original from 1. that's why you have to do it this way.

I don't see another way here.

Reply:
brokers will send an ack over the wire towards the other server. 

if I don't mix the IDs here, I will have to send two values, and make the key a complex object, which will take a hit in performance for this. I think it's worth to make it this way.

Reply:
@gemmellr you concerned about hops on this case?

This is something I was still pending on the work I'm doing here.

I am planning to send the same ID if already had a previous ID set.

Reply:
I am going to add a test with a mirror of a mirror to see how these IDs would behave. 

The idea is that the internal ID on its original node should be kept all the way. I am missing a test for that, and I will take care of this before I set this PR ready to merge.

Reply:
Not concerned about hops no, just the ugly hoop jumping and 'mix then unmix to log' etc and general fragility of the munging, plus bizarre mismatch between what is on the sent message (separate values) and the acks (munged values).

Reply:
the sent and acks was per your request...


I didn't want to separate the list per nodeIDs...

I don't think the duplication of IDs would be an issue really. But I have an idea that may fix it. with some impact in performance of course, but I will try to minimize it.

Reply:
I dont believe I requested that. I do recall questioning why some of the event detail was going in message-annotations and some of it was in the body for the acks and suggesting they be consolidated. I also recall not particularly liking the delivery-annotations usage for sends initially as I thought they might fall foul of their 'doesnt travel with message' nature when message-routing through Dispatch, but it turned out that wasnt an issue.

(It occurs to me in thinking about and typing that, that the new addition of a reliance on the 'mirror id' link property will effectively rule out message-routing based mirroring in any case...or at least break it).

Adding the mirror-id to the annotations which already exist would be one option of not munging the values for acking. Or changing the body, e.g using an array, or a list, or Binary and encoding the detail in full. Or a Data section. Etc.

Reply:
> (It occurs to me in thinking about and typing that, that the new addition of a reliance on the 'mirror id' link property will effectively rule out message-routing based mirroring in any case...or at least break it).

Which to be clear, I think is fine...but should be handled/documented. E.g perhaps the links should be using the capabilities to announce their desired capabilities and detect the peers offered capabilities to request and detect support of this mirroring functionality, so that its absence is detected and fails early. That woudl also stop it being used with old brokers that dont understand the changes made.

Reply:
Passing two variables won't help anyway as it will still need to mix the IDs to store and ack on the list. 


I am rethinking this.  Perhaps use the NodeID and play with multiple lists.  

I will have it ready tomorrow. Will initially keep a separate commit to squash later.

Reply:
Yes, I wasnt suggesting that would be the _only_ change required.

Reply:
I sent a commit as an attempt to remove the mirrorID and just use another value with the node UUID. it will be another Object and another list, but it will work.

right now things are not 100% working as I'm getting some reflection on the test cases (sends happening again after the mirror is sent).. I will fix this tomorrow... 

the last commit still a WIP and a lot of that will still change.. I sent to the PR just a reference of my attempt.

Reply:
I actually got the tests fixed.. I will spend some time cleaning up tomorrow.

but the mixed IDs are gone now.